File: app/src/main/java/opencrowd/hgc/hgcwallet/ui/main/request/RequestListScreen.java
package opencrowd.hgc.hgcwallet.ui.main.request;


import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import com.wealthfront.magellan.BaseScreenView;
import com.wealthfront.magellan.Screen;

import java.util.List;

import opencrowd.hgc.hgcwallet.R;
import opencrowd.hgc.hgcwallet.common.Singleton;
import opencrowd.hgc.hgcwallet.database.DBHelper;
import opencrowd.hgc.hgcwallet.database.account.Account;
import opencrowd.hgc.hgcwallet.database.request.PayRequest;
import opencrowd.hgc.hgcwallet.modals.HGCAccountID;
import opencrowd.hgc.hgcwallet.ui.customviews.TitleBarWrapper;
import opencrowd.hgc.hgcwallet.ui.main.pay.PayScreen;

public class RequestListScreen extends Screen<RequestListView> {

    @Override
    protected RequestListView createView(Context context) {
        return new RequestListView(context);
    }
}

class RequestListView  extends BaseScreenView<RequestListScreen> {

    private List<PayRequest> mRequestList;
    private RecyclerView mRecyclerView;
    private RequestListAdapter mAdapter;


    public RequestListView(Context context) {
        super(context);
        inflate(context, R.layout.view_accout_list_layout,this);
        mRequestList = DBHelper.getAllRequests();
        TitleBarWrapper titleBar = new TitleBarWrapper(findViewById(R.id.titleBar));
        titleBar.setTitle("REQUESTS");
        if (mRequestList.size() > 0) {
            mRecyclerView = (RecyclerView) findViewById(R.id.account_list_recyclerview);
            mRecyclerView.setLayoutManager(new LinearLayoutManager(context));
            mRecyclerView.setItemAnimator(new DefaultItemAnimator());
            mRecyclerView.setHasFixedSize(true);
            mAdapter = new RequestListAdapter(mRequestList);
            mRecyclerView.setAdapter(mAdapter);
            mAdapter.setmRequestListner(new RequestListAdapter.OnPayRequestClick() {
                @Override
                public void onRequestPick(PayRequest toAccount) {
                    // below code should be discuss
                    List<Account> accountList = DBHelper.getAllAccounts();
                    Account fromAccount = accountList.get(0);
                    getScreen().getNavigator().goTo(new PayScreen(fromAccount, toAccount));
                }
            });
        }
    }
}

class RequestListAdapter extends RecyclerView.Adapter<RequestListAdapter.MyViewHolder> {

    private List<PayRequest> mRequestList;

    public RequestListAdapter(List<PayRequest> requestList) {
        this.mRequestList = requestList;
    }

    @NonNull
    @Override
    public MyViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.request_list_row, parent, false);

        return new MyViewHolder(itemView);
    }

    @NonNull


    @Override
    public void onBindViewHolder(@NonNull MyViewHolder holder, int position) {
        final PayRequest payRequest = mRequestList.get(position);
        holder.mTextViewAccountName.setVisibility(View.VISIBLE);
        if(payRequest.getName() != null && !payRequest.getName().isEmpty()) {
            holder.mTextViewAccountName.setText(payRequest.getName());
        }else {
            holder.mTextViewAccountName.setText("UNKNOWN");
        }
        if(payRequest.getNotes() != null && !payRequest.getNotes().isEmpty()) {
            holder.mTextViewNotes.setText(payRequest.getNotes());
            holder.mTextViewNotes.setVisibility(View.VISIBLE);
        }else {
            holder.mTextViewNotes.setVisibility(View.GONE);
        }
        if(payRequest.getAccountId() != null && !payRequest.getAccountId().isEmpty()) {
            holder.mTextViewAccountId.setText(payRequest.getAccountId());
            holder.mTextViewAccountId.setVisibility(View.VISIBLE);
        }else {
            holder.mTextViewAccountId.setVisibility(View.GONE);
        }


        if(payRequest.getImportDate() != null) {
            holder.mTextViewTime.setText(Singleton.INSTANCE.getDateFormat(payRequest.getImportDate()));
            holder.mTextViewTime.setVisibility(View.VISIBLE);
        }else {
            holder.mTextViewTime.setVisibility(View.GONE);
        }

        if (payRequest.getAmount() > 0) {
            holder.mTextViewDollerValue.setText(Singleton.INSTANCE.formatUSD(payRequest.getAmount(),true));
            holder.mTextViewHgcWalletValue.setText(Singleton.INSTANCE.formatHGCShort(payRequest.getAmount(), true));
            holder.mTextViewDollerValue.setVisibility(View.VISIBLE);
            holder.mTextViewHgcWalletValue.setVisibility(View.VISIBLE);
        } else {
            holder.mTextViewDollerValue.setVisibility(View.INVISIBLE);
            holder.mTextViewHgcWalletValue.setVisibility(View.INVISIBLE);
        }

        holder.mButtonPay.setVisibility(View.VISIBLE);

        holder.mButtonPay.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if(mRequestListner != null) {
                    mRequestListner.onRequestPick(payRequest);
                }
            }
        });
    }


    @Override
    public int getItemCount() {
        return mRequestList.size();
    }

    public class MyViewHolder extends RecyclerView.ViewHolder {
        public TextView mTextViewTime,mTextViewAccountName, mTextViewHgcWalletValue, mTextViewAccountId,mTextViewDollerValue,mTextViewNotes;
        public Button mButtonPay;
        public MyViewHolder(@NonNull View view) {
            super(view);
            mTextViewTime = (TextView) view.findViewById(R.id.text_time);
            mTextViewAccountName = (TextView) view.findViewById(R.id.text_account_name);
            mTextViewHgcWalletValue = (TextView) view.findViewById(R.id.hgc_wallet_text);
            mTextViewAccountId = (TextView) view.findViewById(R.id.text_key);
            mTextViewDollerValue = (TextView)view.findViewById(R.id.dollor_text);
            mTextViewNotes = (TextView)view.findViewById(R.id.notes_text);
            mButtonPay = (Button)view.findViewById(R.id.btn_pay);
        }
    }

    public interface OnPayRequestClick {
        public void onRequestPick(PayRequest payRequest);
    }

    OnPayRequestClick mRequestListner;

    public void setmRequestListner(OnPayRequestClick onRequestClick) {
        this.mRequestListner = onRequestClick;
    }
}

File: app/src/main/java/hedera/hgc/hgcwallet/ui/main/request/RequestListScreen.kt
/*
 *
 *  Copyright 2019 Hedera Hashgraph LLC
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */

package hedera.hgc.hgcwallet.ui.main.request


import android.content.Context
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView

import com.wealthfront.magellan.BaseScreenView
import com.wealthfront.magellan.Screen

import hedera.hgc.hgcwallet.R
import hedera.hgc.hgcwallet.common.Singleton
import hedera.hgc.hgcwallet.database.DBHelper
import hedera.hgc.hgcwallet.database.account.Account
import hedera.hgc.hgcwallet.database.request.PayRequest
import hedera.hgc.hgcwallet.ui.customviews.TitleBarWrapper
import hedera.hgc.hgcwallet.ui.main.pay.PayScreen
import hedera.hgc.hgcwallet.ui.main.pay.PayType

class RequestListScreen : Screen<RequestListView>() {

    data class Params(val requestList: List<PayRequest>, val accountList: List<Account>)

    private val params = Params(DBHelper.getAllRequests() ?: listOf(), DBHelper.getAllAccounts())

    override fun createView(context: Context): RequestListView {
        return RequestListView(context, params)
    }

    internal fun onPayRequestForAccount(payRequest: PayRequest) {
        val fromAccount = params.accountList.first()
        navigator?.goTo(PayScreen(fromAccount, payRequest, PayType.Pay))
    }
}

class RequestListView(context: Context, val params: RequestListScreen.Params) : BaseScreenView<RequestListScreen>(context) {

    private var mRecyclerView: RecyclerView? = null

    init {
        View.inflate(context, R.layout.view_accout_list_layout, this)
        TitleBarWrapper(findViewById(R.id.titleBar)).apply {
            setTitle("REQUESTS")
        }

        if (params.requestList.isNotEmpty()) {
            val mAdapter = RequestListAdapter(params.requestList, object : RequestListAdapter.OnPayRequestClick {
                override fun onRequestPick(payRequest: PayRequest) {
                    // below code should be discuss
                    screen?.onPayRequestForAccount(payRequest)
                }
            })

            mRecyclerView = findViewById<RecyclerView>(R.id.account_list_recyclerview)?.apply {
                layoutManager = LinearLayoutManager(context)
                itemAnimator = DefaultItemAnimator()
                setHasFixedSize(true)
                adapter = mAdapter
            }
        }
    }
}

internal class RequestListAdapter(private val requestList: List<PayRequest>, private val requestListener: OnPayRequestClick?) : RecyclerView.Adapter<RequestListAdapter.MyViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyViewHolder {
        val itemView = LayoutInflater.from(parent.context)
                .inflate(R.layout.request_list_row, parent, false)

        return MyViewHolder(itemView)
    }

    override fun onBindViewHolder(holder: MyViewHolder, position: Int) {
        val payRequest = requestList[position]
        holder.mTextViewAccountName.visibility = View.VISIBLE

        holder.mTextViewAccountName.text = if (!payRequest.name.isNullOrEmpty()) payRequest.name else "UNKNOWN"

        if (!payRequest.notes.isNullOrEmpty()) {
            holder.mTextViewNotes.text = payRequest.notes
            holder.mTextViewNotes.visibility = View.VISIBLE
        } else {
            holder.mTextViewNotes.visibility = View.GONE
        }

        if (payRequest.accountId.isNotEmpty()) {
            holder.mTextViewAccountId.text = payRequest.accountId
            holder.mTextViewAccountId.visibility = View.VISIBLE
        } else {
            holder.mTextViewAccountId.visibility = View.GONE
        }


        holder.mTextViewTime.text = Singleton.getDateFormat(payRequest.importDate)
        holder.mTextViewTime.visibility = View.VISIBLE


        if (payRequest.amount > 0) {
            holder.mTextViewDollerValue.text = Singleton.formatUSD(payRequest.amount, true)
            holder.mTextViewHgcWalletValue.text = Singleton.formatHGCShort(payRequest.amount, true)
            holder.mTextViewDollerValue.visibility = View.VISIBLE
            holder.mTextViewHgcWalletValue.visibility = View.VISIBLE
        } else {
            holder.mTextViewDollerValue.visibility = View.INVISIBLE
            holder.mTextViewHgcWalletValue.visibility = View.INVISIBLE
        }

        holder.mButtonPay.visibility = View.VISIBLE

        holder.mButtonPay.setOnClickListener {
            requestListener?.onRequestPick(payRequest)
        }
    }


    override fun getItemCount(): Int {
        return requestList.size
    }

    inner class MyViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        var mTextViewTime: TextView
        var mTextViewAccountName: TextView
        var mTextViewHgcWalletValue: TextView
        var mTextViewAccountId: TextView
        var mTextViewDollerValue: TextView
        var mTextViewNotes: TextView
        var mButtonPay: Button

        init {
            mTextViewTime = view.findViewById<View>(R.id.text_time) as TextView
            mTextViewAccountName = view.findViewById<View>(R.id.text_account_name) as TextView
            mTextViewHgcWalletValue = view.findViewById<View>(R.id.hgc_wallet_text) as TextView
            mTextViewAccountId = view.findViewById<View>(R.id.text_key) as TextView
            mTextViewDollerValue = view.findViewById<View>(R.id.dollor_text) as TextView
            mTextViewNotes = view.findViewById<View>(R.id.notes_text) as TextView
            mButtonPay = view.findViewById<View>(R.id.btn_pay) as Button
        }
    }

    interface OnPayRequestClick {
        fun onRequestPick(payRequest: PayRequest)
    }
}
